### ISSUE TITLE
SSL3_READ_BYTES:tlsv1 alert protocol version

### ISSUE THREAD

**Issue by wholeinsoul** [jazzband/django-push-notifications#493]
django-push-notifications was working with both sandbox and production certificates. But since yesterday it is failing with following errors. No change in our code or our servers. Any ideas, why it has stopped working.

13/Mar/2019 13:52:23] ERROR [django.request:256] Internal Server Error: /shadmin/push_notifications/apnsdevice/ Traceback (most recent call last):

File "/home/docker/env/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
  response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 616, in wrapper
  return self.admin_site.admin_view(view)(*args, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 110, in _wrapped_view
  response = view_func(request, *args, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/django/views/decorators/cache.py", line 57, in _wrapped_view_func
  response = view_func(request, *args, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/django/contrib/admin/sites.py", line 233, in inner
  return view(request, *args, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
  return bound_func(*args, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 110, in _wrapped_view
  response = view_func(request, *args, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
  return func.__get__(self, type(self))(*args2, **kwargs2)
File "/home/docker/env/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1574, in changelist_view
  response = self.response_action(request, queryset=cl.get_queryset(request))
File "/home/docker/env/local/lib/python2.7/site-packages/django/contrib/admin/options.py", line 1333, in response_action
  response = func(self, request, queryset)
File "/home/docker/env/local/lib/python2.7/site-packages/push_notifications/admin.py", line 51, in send_message
  self.send_messages(request, queryset)
File "/home/docker/env/local/lib/python2.7/site-packages/push_notifications/admin.py", line 30, in send_messages
  r = device.send_message("Test single notification")
File "/home/docker/env/local/lib/python2.7/site-packages/push_notifications/models.py", line 93, in send_message
  return apns_send_message(registration_id=self.registration_id, alert=message, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/push_notifications/apns.py", line 209, in apns_send_message
  _apns_send(registration_id, alert, **kwargs)
File "/home/docker/env/local/lib/python2.7/site-packages/push_notifications/apns.py", line 153, in _apns_send
  with closing(_apns_create_socket_to_push()) as socket:
File "/home/docker/env/local/lib/python2.7/site-packages/push_notifications/apns.py", line 58, in _apns_create_socket_to_push
  return _apns_create_socket((SETTINGS["APNS_HOST"], SETTINGS["APNS_PORT"]))
File "/home/docker/env/local/lib/python2.7/site-packages/push_notifications/apns.py", line 52, in _apns_create_socket
  sock.connect(address_tuple)
File "/usr/lib/python2.7/ssl.py", line 433, in connect
  self._real_connect(addr, False)
File "/usr/lib/python2.7/ssl.py", line 423, in _real_connect
  self.do_handshake()
File "/usr/lib/python2.7/ssl.py", line 405, in do_handshake
  self._sslobj.do_handshake()
SSLError: [Errno 1] _ssl.c:510: error:1409442E:SSL routines:SSL3_READ_BYTES:tlsv1 alert protocol version

**Comment by cspinelive** [comment]
same here

**Comment by cspinelive** [comment]
https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/CommunicatingwithAPNs.html

mentions

> Your provider must support TLS 1.2 or higher when connecting to APNs.

We are looking at upgrading openssl on our ubuntu server to see if that helps

**Comment by wholeinsoul** [comment]
@cspinelive : You can try but I doubt it is the problem. I already looked at this document. If you see the revision history of this document, it was last updated on 2018-06-04. And our push notifications were working until day before yesterday.

https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/RevisionHistory.html#//apple_ref/doc/uid/TP40008194-CH99-SW1

**Comment by wholeinsoul** [comment]
FYI: I tried regenerating apple push certificates but still same error.

**Comment by cspinelive** [comment]
just adding more info... we are seeing that connections will work about half the time

**Comment by wholeinsoul** [comment]
> just adding more info... we are seeing that connections will work about half the time

Yep. That is my observation too. It is random but most of the times it is failing.

Did you try upgrading the openssl ?
May be you were right that it could be openssl:

Running:
openssl s_client -connect api.push.apple.com:443 -tls1_2

returns:
XHTTP/2 client preface string missing or corrupt. Hex dump for received bytes: 0aread:errno=0

which is not supported by OpenSSL 1.0.1f 6 Jan 2014 and this is the version installed on my server.

So now I'm also trying to update openssl.

**Comment by cspinelive** [comment]
We're behind the times.  Still using the older binary interface to send the pushes.  We're also using python 2.7.6 on ubuntu 12.04.  Everything I'm seeing on the linux side tells me we've got openssl and tls 1.2, but I'm seeing in the python docs that python might be specifying tls v1 as 1.2 wasn't supported by python until 2.7.9.  Upgrading to a later python version isn't as simple as I'd like because of my ubuntu version.  

We're a week or so away from upgrading our infrastructure to ubuntu 18.04 and python 3.6 so we've got those servers already set up and mostly tested.  I can confirm that the issue is not present on those servers.

So right now as a band aid, and to get me back up and running, I'm going to put in some retry logic when making the connection to apple.  Then in a few weeks remove it once we upgrade to our new infrastructure.

**Comment by cspinelive** [comment]
I too have OpenSSL 1.0.1f 6 Jan 2014

**Comment by cspinelive** [comment]
I might try https://www.miguelvallejo.com/updating-to-openssl-1-0-2g-on-ubuntu-server-12-04-14-04-lts-to-stop-cve-2016-0800-drown-attack/

https://askubuntu.com/questions/429385/upgrade-openssl-on-ubuntu-12-04
indicates that upgrading to f was enough for them.

**Comment by wholeinsoul** [comment]
I upgraded to OpenSSL 1.1.1a  20 Nov 2018 but same results.
I also checked TLS version supported by my python using following and got "TLS 1.3". So not sure what else to try next.

`python -c "import requests; print(requests.get('https://www.howsmyssl.com/a/check', verify=False).json()['tls_version'])"
`
